galaxy_ng/app/migrations/0037_namespace_sync.py

```diff
@@ -0,0 +1,153 @@
+# Generated by Django 3.2.18 on 2023-02-15 13:52
+
+import django.core.validators
+from django.db import migrations
+from django.db.models import Q
+import hashlib
+import json
+
+def calculate_metadata_sha256(metadata):
+    """Calculates the metadata_sha256 from the other metadata fields."""
+    metadata_json = json.dumps(metadata, sort_keys=True).encode("utf-8")
+    hasher = hashlib.sha256(metadata_json)
+    
+    return hasher.hexdigest()
+
+
+def add_pulp_ansible_namespace_metadata_objects(apps, schema_editor):
+    """Migrate galaxy namespaces to pulp ansible namespace metadata content."""
+
+    """
+    for each namespace:
+    - create AnsibleNamespaceMetadata object
+        - calculate the metadata sha
+    - add namespace permissions to pulp namespace obj
+        - or maybe not? should we just enforce the rbac on our
+          namespaces?
+
+    Add the new namespace to the following repos:
+        - published
+        - should these be added to the rest?
+    """
+
+    AnsibleNamespaceMetadata = apps.get_model('ansible', 'AnsibleNamespaceMetadata')
+    AnsibleNamespace = apps.get_model('ansible', 'AnsibleNamespace')
+    Namespace = apps.get_model('galaxy', 'Namespace')
+
+    # ContentType = apps.get_model("contenttypes", "ContentType")
+    # GroupRole = apps.get_model("core", "GroupRole")
+    # old_ns_type = ContentType.objects.get(app_label="galaxy", model="namespace")
+    # new_ns_type = ContentType.objects.get(app_label="ansible", model="ansiblenamespace")
+
+
+    for old_ns in Namespace.objects.all():
+        new_ns = AnsibleNamespace.objects.create(name=old_ns.name)
+        links = {l.name: l.url for l in old_ns.links.all()}
+
+        metadata = {
+            "company": old_ns.company,
+            "email": old_ns.email,
+            "description": old_ns.description,
+            "resources": old_ns.resources,
+            "links": links,
+            "avatar_sha256": None,
+            "avatar_url": old_ns.avatar_url,
+        }
+
+        # skip metadata creation for namespaces with no data.
+        create_metadata = False
+        for f in metadata:
+            if metadata[f] not in ('', dict(), None):
+                create_metadata = True
+                break
+
+        if create_metadata:
+            metadata["name"] = old_ns.name
+            AnsibleNamespaceMetadata.objects.create(
+                **{
+                    **metadata,
+                    "metadata_sha256": calculate_metadata_sha256(metadata),
+                    "namespace": new_ns,
+                    "pulp_type": "ansible.namespace",
+                }
+            )
```

Comment:
```suggestion
            AnsibleNamespaceMetadata.objects.create(namespace=new_ns, **metadata)
```
`pulp_type` is auto added on save and `metadata_sha256` is also auto-calculated before the save. You can remove `calculate_metadata_sha256`

Reply:
I don't think those hooks are called during migrations, but I can try.